activerecord/lib/active_record/relation/predicate_builder.rb

```diff
@@ -149,19 +149,15 @@ def grouping_queries(queries)
       end
 
       def convert_dot_notation_to_hash(attributes)
-        dot_notation = attributes.select do |k, v|
-          k.include?(".") && !v.is_a?(Hash)
-        end
-
-        dot_notation.each_key do |key|
-          table_name, column_name = key.split(".")
-          value = attributes.delete(key)
-          attributes[table_name] ||= {}
-
-          attributes[table_name] = attributes[table_name].merge(column_name => value)
+        attributes.each_with_object({}) do |(key, value), converted|
+          if !value.is_a?(Hash) && (idx = key.rindex("."))
+            table_name, column_name = key[0, idx], key[idx + 1, key.length]
+            converted[table_name] ||= {}
+            converted[table_name] = converted[table_name].merge(column_name => value)
```

Comment:
Any reason why you don't do: 

```suggestion
            converted[table_name][column_name => value]
```

? That would avoid lots of Hash allocations.